### PR TITLE
Increase warning level

### PR DESCRIPTION
**Pull request by stotko**
- Increase warning level for GNU/Clang/Intel compilers by `-Wextra`, but disable noisy `-Wunused-parameter` warning for now
- Make CUDA flags consistent with C++ flags
- Fix newly introduced warnings by the above bump

---
This change is [Reviewable](https://reviewable.io/reviews/intel-isl/open3d/3357)

### CODE REVIEW

**Review by prewettg — APPROVED** [review]
> Reviewed 8 of 9 files at r1.
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/3357)* status: 8 of 9 files reviewed, 1 unresolved discussion (waiting on @errissa, @ssheorey, @stotko, and @yxlao)

___
*[cpp/open3d/geometry/TriangleMesh.cpp, line 1498 at r1](https://reviewable.io/reviews/intel-isl/open3d/3357#-MZTLMZQ-GrcriQMLrG9:-MZTLMZQ-GrcriQMLrGA:bk4jep6) ([raw file](https://github.com/intel-isl/open3d/blob/51b10892a63d0212f60f130c68651c2937cdea98/cpp/open3d/geometry/TriangleMesh.cpp#L1498)):*
> ```cpp
>     std::vector<bool> triangle_mask(triangles_.size(), false);
>     for (auto tidx : triangle_indices) {
>         if (/*tidx >= 0 &&*/ tidx < triangles_.size()) {
> ```

We can just delete this (and below), right? Since size() is unsigned, tidx is always >= 0. No point it commenting it out and future generations wondering why it's still there but not removing it just in case...

File: cpp/CMakeLists.txt [hunk 1]
```diff
@@ -17,54 +17,58 @@ function(open3d_show_and_abort_on_warning trgt)
...
+        string(REPLACE ";" "," DISABLE_NVCC_WARNINGS "${DISABLE_NVCC_WARNINGS}")
 
-        if (CMAKE_CUDA_COMPILER_VERSION VERSION_LESS "10.2")
-            if (MSVC)
-                string(APPEND CUDA_FLAGS " -Xcompiler /W4,/WX")
-            else()
-                string(APPEND CUDA_FLAGS " -Xcompiler -Wall,-Werror")
-            endif()
-        else()
+        set(CUDA_FLAGS "--Werror cross-execution-space-call,deprecated-declarations,reorder")
```
Comment (Dudulle) [inline cpp/CMakeLists.txt:34]: `reorder` should not be set for all platforms as it breaks builds on Windows.

File: cpp/CMakeLists.txt [hunk 2]
```diff
@@ -17,54 +17,58 @@ function(open3d_show_and_abort_on_warning trgt)
...
-            string(APPEND CUDA_FLAGS " --expt-relaxed-constexpr")
-            if (CMAKE_CUDA_COMPILER_VERSION VERSION_LESS "11.2")
-                string(APPEND CUDA_FLAGS " -Xcudafe --diag_suppress=[${DISABLE_NVCC_WARNINGS}]")
-            else()
-                string(APPEND CUDA_FLAGS " --diag-suppress ${DISABLE_NVCC_WARNINGS}")
-            endif()
+
+            string(APPEND CUDA_FLAGS " -Xcompiler /W4,/WX,${CUDA_DISABLE_MSVC_WARNINGS}")
         else()
-            string(APPEND CUDA_FLAGS " --Werror reorder")
```
Comment (Dudulle) [inline cpp/CMakeLists.txt:57]: See previous comment.

**Review by yxlao — APPROVED** [review]
> Reviewed 9 of 9 files at r1.
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/3357)* status: all files reviewed, 1 unresolved discussion (waiting on @errissa, @ssheorey, and @stotko)

**Review by ssheorey — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/3357)* status: all files reviewed, 3 unresolved discussions (waiting on @errissa and @stotko)

___
*[cpp/CMakeLists.txt, line 22 at r1](https://reviewable.io/reviews/intel-isl/open3d/3357#-MZUGxUG5Jq7ouKEKGhf:-MZUGxUG5Jq7ouKEKGhg:b-l9jkks) ([raw file](https://github.com/intel-isl/open3d/blob/51b10892a63d0212f60f130c68651c2937cdea98/cpp/CMakeLists.txt#L22)):*
> ```CMake
>     set(DISABLE_GNU_CLANG_INTEL_WARNINGS
>             -Wno-unused-parameter               # (many places in Open3D code)
>             -Wno-missing-field-initializers     # (visualization/gui/Events.h)
> ```

If this is only one file, we can use`#pragma GCC diagnostic push/pop` inside the file, so there is a reminder for the next person updating the file to fix it.

**Review by stotko — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/intel-isl/open3d/3357)* status: 7 of 14 files reviewed, 3 unresolved discussions (waiting on @Dudulle, @errissa, @prewettg, @ssheorey, and @yxlao)

___
*[cpp/CMakeLists.txt, line 22 at r1](https://reviewable.io/reviews/intel-isl/open3d/3357#-MZUGxUG5Jq7ouKEKGhf:-MZXDco2Bp5UNEhQBP6Y:b-19ngzg) ([raw file](https://github.com/intel-isl/open3d/blob/51b10892a63d0212f60f130c68651c2937cdea98/cpp/CMakeLists.txt#L22)):*
<details><summary><i>Previously, ssheorey wrote…</i></summary><blockquote>

If this is only one file, we can use`#pragma GCC diagnostic push/pop` inside the file, so there is a reminder for the next person updating the file to fix it.
</blockquote></details>

Done by fixing the warning.
___
*[cpp/CMakeLists.txt, line 34 at r1](https://reviewable.io/reviews/intel-isl/open3d/3357#-MZSGEtYWaBz1W8o5rah-r1-34:-MZXECnhBJprA1oPqgnc:b-896fix) ([raw file](https://github.com/intel-isl/open3d/blob/51b10892a63d0212f60f130c68651c2937cdea98/cpp/CMakeLists.txt#L34)):*
<details><summary><i>Previously, Dudulle (Cyril Kaiser) wrote…</i></summary><blockquote>

`reorder` should not be set for all platforms as it breaks builds on Windows.
</blockquote></details>

Done.
___
*[cpp/open3d/geometry/TriangleMesh.cpp, line 1498 at r1](https://reviewable.io/reviews/intel-isl/open3d/3357#-MZTLMZQ-GrcriQMLrG9:-MZXDlMh0q-0SKMYZA2n:b-896fix) ([raw file](https://github.com/intel-isl/open3d/blob/51b10892a63d0212f60f130c68651c2937cdea98/cpp/open3d/geometry/TriangleMesh.cpp#L1498)):*
<details><summary><i>Previously, prewettg wrote…</i></summary><blockquote>

We can just delete this (and below), right? Since size() is unsigned, tidx is always >= 0. No point it commenting it out and future generations wondering why it's still there but not removing it just in case...
</blockquote></details>

Done.

File: cpp/CMakeLists.txt (hunk 1 above)
Comment (stotko) [reply cpp/CMakeLists.txt]: Fixed, thanks.

**Pull request merged**